### PR TITLE
Demo Nodes with Async entities

### PR DESCRIPTION
**Pull request by jparisu**
Requirements:

- #75

### CODE REVIEW

File: amlip_demo_nodes/README.md [hunk 1]
```diff
@@ -3,18 +3,40 @@
 This is a package with a demo on AML-IP node execution.
```
Comment (irenebm) [inline amlip_demo_nodes/README.md:3]:
```suggestion
This is a package containing a demo of AML-IP node execution.
```

File: amlip_demo_nodes/README.md [hunk 2]
```diff
@@ -3,18 +3,40 @@
...
+
+---
+
 ## Demo
 
-This intends to recreate a simple real scenario where a user with a C++ based *Computing Node* waits and answers
-tasks from a Python implementation of a *Main Node*.
+This intends to recreate a simple real scenario where *Computing Nodes* waits and answers
+tasks from *Main Nodes*.
 These nodes can be run from anywhere in the LAN, and as much of each kind as desired, and the workload
```
Comment (irenebm) [inline amlip_demo_nodes/README.md:29]:
```suggestion
These nodes can be run from anywhere in the LAN and as much of each kind as desired. The workload
```

File: amlip_demo_nodes/README.md [hunk 3]
```diff
@@ -28,37 +50,39 @@ However, to make the user experience more realistic, the calculation of this to-
 ### Main Node
 
 This is a **Python** application that sends jobs in the format of strings.
-From each string given it creates a job, send it and awaits for a computing to receive the solution.
-There are 2 ways to introduce strings to send jobs:
+From each string given it creates a job and send it with a configurable time elapsed between job sends.
```
Comment (irenebm) [inline amlip_demo_nodes/README.md:53]:
```suggestion
From each string given it creates a job and sends it with a configurable time elapsed between job sends.
```

File: amlip_demo_nodes/README.md [hunk 4]
```diff
@@ -28,37 +50,39 @@ However, to make the user experience more realistic, the calculation of this to-
...
 ```
 
-Its implementation can be seen in `main_node.py` file.
+For further information regarding executable arguments, use `--help` argument.
 
 ### Computing Node
 
-This is a **C++** application that waits for a job to arrive from a Main node and returns the solution.
-It receives an integer argument to set the number of tasks that this node must wait before closing.
+This is a **Python** application that waits for a job to arrive from a Main node and returns the solution.
```
Comment (irenebm) [inline amlip_demo_nodes/README.md:69]:
```suggestion
This is a **Python** application that waits for a job to arrive from a *Main node* and returns the solution.
```

File: amlip_demo_nodes/README.md [hunk 5]
```diff
@@ -28,37 +50,39 @@ However, to make the user experience more realistic, the calculation of this to-
...
 
-Its implementation can be seen in `computing_node.cpp` file.
+For further information regarding executable arguments, use `--help` argument.
 
 ---
 
-## Build dependencies
+## Build Project
+
+In order to build the following project an be able to operate with this demo Nodes, follow these instructions:
```
Comment (irenebm) [inline amlip_demo_nodes/README.md:85]:
```suggestion
In order to build the following project and be able to operate with this demo Nodes, follow these instructions:
```

File: amlip_demo_nodes/README.md [hunk 6]
```diff
@@ -68,12 +92,45 @@ colcon build --packages-up-to amlip_demo_nodes
 source install/setup.bash
 ```
 
-### Execute nodes
+---
+
+## Real Use Case
+
+In order to adapt these nodes for a real Use case (substitute the string transformations for real calculations) look in files for comment `ALGEBRAIC` and substitute with own code.
```
Comment (irenebm) [inline amlip_demo_nodes/README.md:99]:
```suggestion
In order to adapt these nodes for a real Use case (substitute the string transformations for real calculations), look in files for comment `ALGEBRAIC` and substitute with own code.
```

File: amlip_demo_nodes/README.md [hunk 7]
```diff
@@ -68,12 +92,45 @@ colcon build --packages-up-to amlip_demo_nodes
...
+---
+
+## Synchronous Nodes simulated
+
+There are some demos under `sync` directory that contains old examples on Synchronous Nodes.
+
+### Sync Main Node
+
+This is a **Python** application that sends jobs in the format of strings.
+From each string given it creates a job, send it and awaits for a computing to receive the solution.
```
Comment (irenebm) [inline amlip_demo_nodes/README.md:110]:
```suggestion
From each string given it creates a job, sends it and awaits for a computing to receive the solution.
```

File: amlip_demo_nodes/README.md [hunk 8]
```diff
@@ -68,12 +92,45 @@ colcon build --packages-up-to amlip_demo_nodes
...
+There are 2 ways to introduce strings to send jobs:
+
+1. Each argument to the python command will be converted to a string and sent as a job.
```
Comment (irenebm) [inline amlip_demo_nodes/README.md:113]:
```suggestion
1. Each argument of the Python command is converted to a string and sent as a job.
```

File: amlip_demo_nodes/README.md [hunk 9]
```diff
@@ -68,12 +92,45 @@ colcon build --packages-up-to amlip_demo_nodes
...
+2. If no arguments are given, it expects user to enter strings by keyboard input. Entering an empty string will close the execution.
```
Comment (irenebm) [inline amlip_demo_nodes/README.md:114]:
```suggestion
2. If no arguments are given, it expects user to enter strings by keyboard input. Entering an empty string will terminate the execution.
```

File: amlip_demo_nodes/README.md [hunk 10]
```diff
@@ -28,37 +50,39 @@ However, to make the user experience more realistic, the calculation of this to-
 ### Main Node
 
 This is a **Python** application that sends jobs in the format of strings.
-From each string given it creates a job, send it and awaits for a computing to receive the solution.
-There are 2 ways to introduce strings to send jobs:
+From each string given it creates a job and send it with a configurable time elapsed between job sends.
+For each solution received, it prints the solution.
+It waits until all jobs have been answered.
+Its implementation can be seen in [main_node.py](main_node.py) file.
```
Comment (irenebm) [inline amlip_demo_nodes/README.md:56]:
```suggestion

Its implementation can be seen in [main_node.py](main_node.py) file.
```

File: amlip_demo_nodes/README.md [hunk 11]
```diff
@@ -68,12 +92,45 @@ colcon build --packages-up-to amlip_demo_nodes
...
+
+This is a **C++** application that waits for a job to arrive from a Main node and returns the solution.
+It receives an integer argument to set the number of tasks that this node must wait before closing.
+
+```sh
+# To answer 2 jobs
 ./install/amlip_demo_nodes/bin/computing_node 2
 ```
+
+Its implementation can be seen in `computing_node.cpp` file.
```
Comment (irenebm) [inline amlip_demo_nodes/README.md:136]:
```suggestion
Its implementation can be seen in [sync_computing_node.cpp](sync/sync_computing_node.cpp) file.
```

File: amlip_demo_nodes/README.md [hunk 12]
```diff
@@ -68,12 +92,45 @@ colcon build --packages-up-to amlip_demo_nodes
...
-# To execute Main Node to send 2 jobs
+# To send 2 jobs
 python3 ./install/amlip_demo_nodes/bin/main_node.py first_job "second job"
 
-# To execute Computing Node to answer 2 jobs
+# To waits for keyboard input
+python3 ./install/amlip_demo_nodes/bin/main_node.py
+```
+
+Its implementation can be seen in `main_node.py` file.
```
Comment (irenebm) [inline amlip_demo_nodes/README.md:124]:
```suggestion
Its implementation can be seen in  [sync_main_node.py](sync/sync_main_node.py) file.
```

File: amlip_demo_nodes/computing_node.py [hunk 13]
```diff
@@ -0,0 +1,138 @@
+# Copyright 2022 Proyectos y Sistemas de Mantenimiento SL (eProsima).
```
[the remark this answers is not in the document]
Comment (jparisu) [reply amlip_demo_nodes/computing_node.py:1]: The on in sync/sync_main_node.py is a moved file, not new one

File: amlip_demo_nodes/README.md [hunk 14]
```diff
@@ -1,20 +1,47 @@
...
 
 ## Demo
 
-This intends to recreate a simple real scenario where a user with a C++ based *Computing Node* waits and answers
-tasks from a Python implementation of a *Main Node*.
-These nodes can be run from anywhere in the LAN, and as much of each kind as desired, and the workload
-of each *Main* will be distributed along the *Computing* available.
+This intends to recreate a simple real scenario where *Computing Nodes* waits and answers tasks from *Main Nodes*.
+These nodes can be run from anywhere in the LAN and as much of each kind as desired.
+Th workload of each *Main* will be distributed along the *Computing* available.
```
Comment (irenebm) [inline amlip_demo_nodes/README.md:35]:
```suggestion
The workload of each *Main* will be distributed along the *Computing* available.
```

File: amlip_demo_nodes/sync/sync_main_node.py [hunk 15]
```diff
@@ -0,0 +1,59 @@
+# Copyright 2022 Proyectos y Sistemas de Mantenimiento SL (eProsima).
```
Comment (irenebm) [inline amlip_demo_nodes/sync/sync_main_node.py:1]:
```suggestion
# Copyright 2023 Proyectos y Sistemas de Mantenimiento SL (eProsima).
```

Comment (jparisu) [reply amlip_demo_nodes/sync/sync_main_node.py:1]: This is not a new file, but a moved one:
https://github.com/eProsima/AML-IP/pull/77#discussion_r1221317673

**Review by irenebm — APPROVED** [review]
LGTM

**Pull request merged**